core/src/toga/hardware/geolocation.py

```diff
@@ -0,0 +1,200 @@
+        cannot be compared directly.
+
+        :returns: An asynchronous result; when awaited, returns True if the app has
+            permission to capture the user's a geolocation while running in the
+            background; False otherwise.
+        :raises PermissionError: If the app has not already requested and received
+            permission to use geolocation services.
+        """
+        result = PermissionResult(None)
+        if not self.has_permission:
+            result.set_exception(
+                PermissionError(
+                    "Cannot ask for background geolocation permission "
+                    "before confirming foreground geolocation permission."
+                )
+            )
+        elif has_background_permission := self.has_background_permission:
+            result.set_result(has_background_permission)
+        else:
+            self._impl.request_background_permission(result)
+
+        return result
+
+    @property
+    def on_change(self) -> OnLocationChangeHandler:
+        """The handler to invoke when an update to the user's location is available."""
+        return self._on_change
+
+    @on_change.setter
+    def on_change(self, handler):
+        self._on_change = wrapped_handler(self, handler)
+
+    def start(self):
+        """Start monitoring the user's location for changes.
+
+        An :any:`on_change` callback will be generated then when the user's location
+        changes.
+
+        :raises PermissionError: If the app has not requested and received permission to
+            use geolocation services.
+        """
+        if self.has_permission:
+            self._impl.start()
+        else:
+            raise PermissionError(
+                "App does not have permission to use geolocation services"
+            )
+
+    def stop(self):
+        """Stop monitoring the user's location.
+
+        :raises PermissionError: If the app has not requested and received permission to
+            use geolocation services.
+        """
+        if self.has_permission:
+            self._impl.stop()
+        else:
+            raise PermissionError(
+                "App does not have permission to use geolocation services"
+            )
+
+    def current_location(self) -> LocationResult:
+        """Obtain the user's current location using the geolocation service.
+
+        If the app hasn't requested and received permission to use geolocation, a
+        :any:`PermissionError` will be raised.
+
+        **This is an asynchronous method**. If you call this method in a synchronous
+        context, it will start the process of requesting the user's location, but will
+        return *immediately*. The return value can be awaited in an asynchronous
+        context, but cannot be compared directly.
+
+        If an :any:`on_change` handler is installed, requesting the current location
+        will also cause that handler to be invoked.
```

Comment:
I agree it's an odd choice; but it's somewhat forced by an annoying implementation quirk. 

On macOS and iOS, you don't explicitly ask for the current location - you can either get the currently cached value (which may not exist), or you request a single location update. There's no "request ID" or other identifier that you can associate with a current location request, so there's no way to differentiate the "source" of an update as being a once-off request, or an ongoing tracking request.

This isn't an issue on Android, as the handler for the two types of update are different; but it's easy to make the behavior consistent. 

The options here (as I see them) are:
1. Generate a change event when current_location is satisfied
2. Don't generate a change event if *any* current_location is being satisfied by an update
3. Modify the API to a passive "request a location update" that expects the user to have a change handler in place to receive the result.

(3) is closer aligned to the macOS/iOS implementation, but seems a bit restrictive as it doesn't let you explicitly request a location and then perform logic on the return value that is different from "I've got a new location" logic. 

(2) would potentially work for macOS/iOS - but it means that if you manually request an update, *and* you've got location tracking enabled, you could potentially miss a location update of interest. I guess having location tracking enabled means you're likely to get an update in another couple of seconds; but if the user stops moving just as they request and update, that might not happen. It's an edge case, but not an inconceivable one.

I ended up going with (1), on the basis that it's not *untrue* - if you request a current location, you *do* have an updated location; and it also allows for you to use the API as if it were (3)-style API (i.e., just install an on_change handler to update a map when *any* change occurs). 

I think I can slightly optimise this by modifying the implementation on macOS/iOS so that a change event is only generated if tracking is enabled. That means we'd only see the on_change update if you're actually tracking location, in which case it becomes one more update callback than you were expecting, but you won't get an on_change notification when you're not location tracking. This would let us remove the explicit on_change that has been added to ensure consistency;  and then modify the documentation you've referenced here to say that requesting the current location *may* cause the handler to be invoked, making the macOS/iOS behavior an implementation quirk, rather than anything guaranteed.